dask/array/core.py

```diff
@@ -2801,6 +2801,8 @@ def concatenate(seq, axis=0, allow_unknown_chunksizes=False):
     --------
     stack
     """
+    seq = [a for a in seq if a.size]
```

Comment:
I guess you and I both found that coercing things into a numpy array is expensive. Is the dask policy that "we don't accept lists or tuples as ndarrays"

Reply:
Honestly I'm not sure I understand the failure here. We access `.shape` on all of these a little later. So something else is off.

In any event, am thinking we should probably move this after the error checks anyways, which might be part of the fix itself.

Reply:
Yeah, I agree we should go ahead and fix that issue as well. Thanks for mentioning that.

Reply:
honestly, I'm all for just telling people to ensure their things are cast. After spending hours on improving the performance of block, just to know that the performance is limited because `asarray` is slow makes it much less rewarding.

Reply:
The reason I say this, is because by the time any of these fixes get done, python 2.7 will be dropped, and most people will be on python 3.5/3.6. At which point, you can just let MyPy loose

Reply:
If you are subscribed to the numpy mailing list, they agree that they made a mistake with the matrix subclass meaning that they were forced to cast everything to a numpy array in all their functions.

I think it is reasonable to say that dask functions accept: "dask.arrays" other things that  expose a reasonable numpy-like interface.

You can choose to use more pure-python methods like `len`, but honestly, it isn't worth it.

Reply:
Understandable. Not sure if you saw, but opened issue ( https://github.com/dask/dask/issues/4169 ) to discuss changing the default behavior of `asarray`, `from_array`, etc. w.r.t. hashing.

Reply:
Marking as resolved (at least for the context of this PR). Hopefully we can address this reasonably in one of the other issues about improving `asarray`.